Add BodyPix ResNet model

tylerzhu-github opened the pull request:
This PR contains changes for BodyPix.js 2.0 release.

DONE:
* Adds ResNet (outputStride=32 and 16 are supported) to BodyPix library.
* Adds multi-person pose, segmentation, and body part support.
* Switches to use GraphModel.
* Switches the MobileNet from LayerModel to GraphModel.
* Adds configuration validation.
* Updates person & body part segmentation decoding logic.
* Updates the camera demo UI to support switching between MobileNet and ResNet.
* Image/video are padded and resized to a square shape using the closes accepted size.
* `estimateSegmentation` methods performs semantic segmentation (and single-person pose).
* `estimateMultiPersonSegmentationByInstance` performs instance segmentation (and multi-person pose).

Please see the README for more details!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-models/280)
<!-- Reviewable:end -->

tylerzhu-github commented:
> Hey just checked this out, super cool this is possible!
> 
> Couple questions:
> 
> It seems like a lot of the code was duplicated from PoseNet - do you think this functionality makes more sense within the PoseNet repo, as just another possible output from the PoseNet ResNet model?
> 
> Will this new model also support multi-person body-party segmentation (not full person segmentation) and how does that fit into the long range offsets?
> 
> Will the MobileNet model also support this new multi-person segmentation?
> 
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-:-LkyoqdNAeJWkhbK8Ibd:bfjlb11)_ status: 0 of 1 approvals obtained (waiting on @dsmilkov)

Hi Dan, thanks for reviewing the draft version.

Please find my answer below:

1. The decoding for instance segmentation requires decoded pose. This is stated in the PersonLab paper too. We could make BodyPix depend on PoseNet, so it can resuse some of the decoding TypeScript logic.

2. Yes it will support multi-person body part segmentation. The decoding is straight-forward since one can just apply the instance mask to crop out a single person's body part segmentation from the body part segmentation prediction for all people. I will finish implementing it this week.

3. Yes the new MobileNet will also support multi-person body part segmentation. I will make sure that the old MobileNet single person model is also converted.

tylerzhu-github commented:
> The new bodypix models now can do everything posenet does, but with more features, correct? If thats the case, what would be the reason to use posenet model instead of bodypix? Are there accuracy differences?
PoseNet will be specialized at sparse keypoints and BodyPix will be specialized toward dense prediction. Currently BodyPix uses sparse pose as a way to decoding the output. 

> If there is no reason to, maybe it's worth long term considering merging these two models into one repo (a much bigger change but just a thought).
Yes once we see a strong convergence toward a particular representation then can consider create a unified model.

tylerzhu-github commented:
Thanks for the detailed suggestions and feedback Dan! I will make sure to address all the points today.

tylerzhu-github commented:
The PR is mostly done!
Feel free to take a stab at reviewing and comments on it!
Thanks

startupgurukul commented:
Looking forward to bodypix 2.0 :)

tylerzhu-github commented:
Thanks for reviewing the PR and I am addressing all the comments.

oveddan commented:
___
*[body-pix/README.md, line 587 at r12](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpOqgUd8tsT1GuBJZ2o:-Lpi8SQK6PHlhhtedUUy:bz3td2) ([raw file](https://github.com/tensorflow/tfjs-models/blob/31c1a9cd9a42ce29019678e98d7d931a0974550f/body-pix/README.md#L587)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Any response here?
</blockquote></details>

I take partial responsibility for this happening, suggesting we leave the existing method of `toMaskImageData` and `toColoredPartImageData` to prevent the api from being a breaking change, because it wasn't entirely clear how the api changes from v1 to v2.  But it seems like from now on the new demo only uses the `multi` methods - is that correct?   Maybe we can remove the `single` methods.


<!-- Sent from Reviewable.io -->

tylerzhu-github commented:
Hi Daniel,  thanks for the reminder!
I will make sure I finished them by end of day today. I am travelling to SF this week and got slightly delayed.

dsmilkov commented:
No worries Tyler. This is a big PR so it's normal for this to take a while to polish - but I'm very excited when we get this out there so all this polishing will pay off!

oveddan commented:
___
*[body-pix/src/multi_person/decode_multiple_masks.ts, line 203 at r20](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpjAG3V6enQCis5-Wh6:-LpjC2s672lMmEMptAGi:bp7c10z) ([raw file](https://github.com/tensorflow/tfjs-models/blob/ae660ae5c0b306f550112841b16dfb347c734795/body-pix/src/multi_person/decode_multiple_masks.ts#L203)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Best to not call `dataSync` - but  instead call `data` multiple times and then waiting for all the datas to resolve.  Same with below 
</blockquote></details>

Here's how it can be written using `data`:

```
export function decodeMultipleMasksGPU(
    segmentation: tf.Tensor, longOffsets: tf.Tensor, poses: Pose[],
    height: number, width: number, stride: number,
    [inHeight, inWidth]: [number, number],
    [[padT, padB], [padL, padR]]: [[number, number], [number, number]],
    minPoseScore = 0.2, refineSteps = 8, minKeypointScore = 0.3,
    maxNumPeople = 10): Array<Promise<PersonSegmentation>> {
  // Filter out poses with smaller score.
  const posesAboveScore = poses.filter(pose => pose.score >= minPoseScore);

  const masksTensor = decodeMultipleMasksTensorGPU(
      segmentation, longOffsets, posesAboveScore, height, width, stride,
      [inHeight, inWidth], [[padT, padB], [padL, padR]], refineSteps,
      minKeypointScore, maxNumPeople);

  return posesAboveScore.map(async (pose, i) => {
    const data =
        (await toPersonKSegmentation(masksTensor, i).data()) as Uint8Array;

    return {height, width, data, pose};
  });
}
```

Calling dataSync causes the drawing thread to be blocked.  If you write it like this, you can 
then invoke it with:

const personMasks = await Promise.all(decodeMultipleMasksTensorGPU(...));


<!-- Sent from Reviewable.io -->

oveddan commented:
___
*[body-pix/src/multi_person/decode_multiple_masks.ts, line 231 at r20](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpjCOWy61dGmYQkNO0V:-LpjCOWy61dGmYQkNO0W:b-mx2vd5) ([raw file](https://github.com/tensorflow/tfjs-models/blob/ae660ae5c0b306f550112841b16dfb347c734795/body-pix/src/multi_person/decode_multiple_masks.ts#L231)):*
> ```XML
>       height,
>       width,
>       data: toPersonKPartSegmentation(masksTensor, partSegmentaion, k)
> ```

Same with above for calling `data` instead of `dataSync`


<!-- Sent from Reviewable.io -->

oveddan commented:
___
*[body-pix/src/multi_person/decode_multiple_masks.ts, line 51 at r20](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpjCd6dEq1D3dAcOh91:-LpjCd6dEq1D3dAcOh92:b-bvoa3p) ([raw file](https://github.com/tensorflow/tfjs-models/blob/ae660ae5c0b306f550112841b16dfb347c734795/body-pix/src/multi_person/decode_multiple_masks.ts#L51)):*
> ```XML
>     [inHeight, inWidth]: [number, number],
>     [[padT, padB], [padL, padR]]: [[number, number], [number, number]],
>     refineSteps: number, minKptScore: number, maxNumPeople: number): tf.Tensor {
> ```

Can you add the dimensionality of the Tensor to the return type?


<!-- Sent from Reviewable.io -->

oveddan commented:
___
*[body-pix/src/multi_person/decode_multiple_masks.ts, line 203 at r20](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpjAG3V6enQCis5-Wh6:-LpjGp9F2lYmma2_CKif:b-2ozq4a) ([raw file](https://github.com/tensorflow/tfjs-models/blob/ae660ae5c0b306f550112841b16dfb347c734795/body-pix/src/multi_person/decode_multiple_masks.ts#L203)):*
<details><summary><i>Previously, oveddan (Dan Oved) wrote…</i></summary><blockquote>

Here's how it can be written using `data`:

```
export function decodeMultipleMasksGPU(
    segmentation: tf.Tensor, longOffsets: tf.Tensor, poses: Pose[],
    height: number, width: number, stride: number,
    [inHeight, inWidth]: [number, number],
    [[padT, padB], [padL, padR]]: [[number, number], [number, number]],
    minPoseScore = 0.2, refineSteps = 8, minKeypointScore = 0.3,
    maxNumPeople = 10): Array<Promise<PersonSegmentation>> {
  // Filter out poses with smaller score.
  const posesAboveScore = poses.filter(pose => pose.score >= minPoseScore);

  const masksTensor = decodeMultipleMasksTensorGPU(
      segmentation, longOffsets, posesAboveScore, height, width, stride,
      [inHeight, inWidth], [[padT, padB], [padL, padR]], refineSteps,
      minKeypointScore, maxNumPeople);

  return posesAboveScore.map(async (pose, i) => {
    const data =
        (await toPersonKSegmentation(masksTensor, i).data()) as Uint8Array;

    return {height, width, data, pose};
  });
}
```

Calling dataSync causes the drawing thread to be blocked.  If you write it like this, you can 
then invoke it with:

const personMasks = await Promise.all(decodeMultipleMasksTensorGPU(...));

</blockquote></details>

Actually, this is better:

```typescript
export async function decodeMultipleMasksGPU(
    segmentation: tf.Tensor, longOffsets: tf.Tensor, poses: Pose[],
    height: number, width: number, stride: number,
    [inHeight, inWidth]: [number, number],
    [[padT, padB], [padL, padR]]: [[number, number], [number, number]],
    minPoseScore = 0.2, refineSteps = 8, minKeypointScore = 0.3,
    maxNumPeople = 10): Promise<PersonSegmentation[]> {
  // Filter out poses with smaller score.
  const posesAboveScore = poses.filter(pose => pose.score >= minPoseScore);

  const personKSegmentations = tf.tidy(() => {
    const masksTensor = decodeMultipleMasksTensorGPU(
        segmentation, longOffsets, posesAboveScore, height, width, stride,
        [inHeight, inWidth], [[padT, padB], [padL, padR]], refineSteps,
        minKeypointScore, maxNumPeople);

    return posesAboveScore.map(
        (pose, k) => ({mask: toPersonKSegmentation(masksTensor, k), pose}));
  });

  const personKSegmentationsData = await Promise.all(personKSegmentations.map(
      async ({pose, mask}) =>
          ({data: await mask.data() as Uint8Array, pose, width, height})));

  personKSegmentations.forEach(({mask}) => mask.dispose());

  return personKSegmentationsData;
}
```

as it takes care of disposing everything from within that function


<!-- Sent from Reviewable.io -->

oveddan commented:
___
*[body-pix/src/multi_person/decode_multiple_masks.ts, line 46 at r20](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpjJplbAY__EOgZ214O:-LpjJplc2xNZUTmLFYF7:b-2evmbf) ([raw file](https://github.com/tensorflow/tfjs-models/blob/ae660ae5c0b306f550112841b16dfb347c734795/body-pix/src/multi_person/decode_multiple_masks.ts#L46)):*
> ```XML
> }
> 
> function decodeMultipleMasksTensorGPU(
> ```

Can you export this to the public api? I, for one, would use it.  My use case is that I just want to get a single segmentation map for the entire image (not segmentation per person) and it would be great if I could call this method in my own custom decoding.

Same for above with `toPersonK` etc....I don't think it needs to  be added to the readme for now.


<!-- Sent from Reviewable.io -->

tylerzhu-github commented:
Thanks Dan! Your PR looks amazing!
I reviewed it and left a couple minor comments.
Let me know when you addressed them and ready to be merged.
I will focus on README.md files for now, so we will not have merge conflict when you sync/pull.

tylerzhu-github commented:
Will address the remaining comments tomorrow. Thanks for the constructive feedbacks.

tylerzhu-github commented:
I have addressed all the comments and feel free to take another pass.
Quick question? Since the BodyPix 2.0 support multi-person should we feature the multi-person GIF at the beginning of the README too? I would like to keep Dan's GIF but add the one with multi-person video.

startupgurukul commented:
Any Chance this can be tried now (before commit)? 

The npm 2.0 version does not exist as yet, please assist
https://cdn.jsdelivr.net/npm/@tensorflow-models/body-pix@2.0.0

oveddan commented:
There are some api changes being implemented before it is released. You can
check out the branch locally and run the demo

On Wed, Oct 2, 2019 at 11:15 PM startupgurukul <notifications@github.com>
wrote:

> Any Chance this can be tried now (before commit)?
>
> The npm 2.0 version does not exist as yet, please assist
> https://cdn.jsdelivr.net/npm/@tensorflow-models/body-pix@2.0.0
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tfjs-models/pull/280?email_source=notifications&email_token=AAGZW2Z2MU4TFJTRCKX4SB3QMVPURA5CNFSM4IG6HKI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAG2VUI#issuecomment-537766609>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGZW2ZEUVDLCPBQTDKQJO3QMVPURANCNFSM4IG6HKIQ>
> .
>
-- 
Sent from Gmail Mobile

startupgurukul commented:
Thank you for the reply
the demos seem to be using bodypix 1.1.0 
    "@tensorflow-models/body-pix": "1.1.0",
https://github.com/tylerzhu-github/tfjs-models/blob/bodypix-resnet/body-pix/demos/package.json

will point to../src and gibe it a try.

Thank you once again

tylerzhu-github commented:
Thanks all for the careful review! I am addressing some of the blocking issues (lint) and READEME update. I will open a bug to track the remaining non-blocking issues.

tylerzhu-github commented:
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/280)_ status: 0 of 1 approvals obtained (waiting on @dsmilkov, @nsthorat, and @oveddan)
> 
> _[body-pix/src/body_pix_test.ts, line 105 at r12](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpOyd0_2TvFCBJ2D3_T:-LqT8vHy8LbhHzTDN8Dw:b4c222r) ([raw file](https://github.com/tensorflow/tfjs-models/blob/31c1a9cd9a42ce29019678e98d7d931a0974550f/body-pix/src/body_pix_test.ts#L105)):_
> 
> Previously, tylerzhu-github (tylerzhu-github) wrote…
> > Thanks for the suggestion. I have created an issue.
> > [tensorflow/tfjs#2153](https://github.com/tensorflow/tfjs/issues/2153)
> 
> Here is my PR:
> [tylerzhu-github#2](https://github.com/tylerzhu-github/tfjs-models/pull/2)
> 
> In it:
> 
> * refactored bodypix to not use datasync in segmentation decoding methods
> * Added back cpu based ops from before, in case webgl is not the backend, and dynamically switching based on the backend.
> * In the demo, added an option to pass, in the query string: backend=cpu to test cpu based decoding.
> * In the demo, defaulting to use MobileNet.
> * Refactored the estimate segmentation methods to have separate activation functions, and be more concise/readable
> * Added some tests around valid resolution
> * Renamed mask decoding methods and file to better imply that they relate to poses, and exported those methods so that they can be used for people that want certain functionality

Great! I have added your items into the Umbrella entry. The dataSync() --> data() has been addressed now in this PR based on the suggested code changes.

oveddan commented:
___
*[body-pix/src/body_pix_test.ts, line 105 at r12](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpOyd0_2TvFCBJ2D3_T:-LqYBqAYBbBsE6_8W5Kt:b1dsd3g) ([raw file](https://github.com/tensorflow/tfjs-models/blob/31c1a9cd9a42ce29019678e98d7d931a0974550f/body-pix/src/body_pix_test.ts#L105)):*
<details><summary><i>Previously, tylerzhu-github (tylerzhu-github) wrote…</i></summary><blockquote>

Great! I have added your items into the Umbrella entry. The dataSync() --> data() has been addressed now in this PR based on the suggested code changes. 
</blockquote></details>

Why not just review and merge the PR that took care of this?   That branch will be more challenging to merge now as of these changes that were duplicated in the main branch.


<!-- Sent from Reviewable.io -->

tylerzhu-github commented:
> _[body-pix/src/body_pix_test.ts, line 105 at r12](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LpOyd0_2TvFCBJ2D3_T:-LqYBqAYBbBsE6_8W5Kt:b1dsd3g) ([raw file](https://github.com/tensorflow/tfjs-models/blob/31c1a9cd9a42ce29019678e98d7d931a0974550f/body-pix/src/body_pix_test.ts#L105)):_
> 
> Previously, tylerzhu-github (tylerzhu-github) wrote…
> Why not just review and merge the PR that took care of this? That branch will be more challenging to merge now as of these changes that were duplicated in the main branch.
I am happy to merge your PR or cherry pick items from your PR. Currently your PR is out of sync with the branch and the PR was closed. I think we can either merge your PR afterwards or create a new branch with the changes cherry-picked.

tylerzhu-github commented:
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/280)_ status: 0 of 1 approvals obtained (waiting on @dsmilkov, @nsthorat, @oveddan, and @tylerzhu-github)
> 
> _[a discussion](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LqaMnjr0H0YO2KqOSMr:-LqaMnjs1WwKu6JBilKt:b9dvglg) (no related file):_
> This users of this PR would benefit greatly from a list of all the things that changed from bodypix 1.0 to 2.0 - things that are only clear to me after digging through the code and using it for a while, and discussing with the team the inner workings of this new model.
> 
> It would be super beneficial if in the PR a description of all the changes to the API are listed, as the current list seems out of date. It should include things such as:
> 
> * Image is no longer automatically resized to a specific dimension, but to the desired input resolution
> * estimateSegmentation methods now also perform pose estimation and return a single pose
> * estimateSegmentation methods capture masks of multiple people (or better worded to explain that it gets masks of everyone together)
> * ResNet model is added without output strides 16 and 32
> 
> Tyler, could you please add a list like this to the PR description? It would really show how awesome the work is that went into this.
> 
> Also, I'm of the strong belief that this PR would be a lot easier to review in the future, look back on, and potentially revert if it didn't contain all of the code that was copied and pasted from posenet, but rather [integrated the PR that just imports that code from PoseNet](https://github.com/tylerzhu-github/tfjs-models/pull/1); This would reduce the PR size by about half. So I will fix that PR with the latest changes, and once it's fixed it would be best if that was integrated into this branch before it was merged.

I have added the suggested items to the PR description. 
Re: share utilities, I have put the suggested item in the Umbrella issue that tracks all following up PRs.

tylerzhu-github commented:
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/280)_ status: 0 of 1 approvals obtained (waiting on @dsmilkov, @nsthorat, @oveddan, and @tylerzhu-github)
> 
> _[a discussion](https://reviewable.io/reviews/tensorflow/tfjs-models/280#-LqaMnjr0H0YO2KqOSMr:-LqagtiT7fYGziaACvfB:b-s15xny) (no related file):_
> 
> Previously, oveddan (Dan Oved) wrote…
> For reference, the PR that took care of this was opened on Aug 22, there has been plenty of time to review/reject it and I would have gladly integrated feedback

Thanks Dan for the detailed suggestions. See reply here for this comment and the the previous comment. I think there was changes in terms of "part" and "keypoint" naming. Also as you mentioned, the required functionality and utilities were not fully exposed publicly back then. To get the release going steadily, let's follow up all the items in the issue I created. I will attach your PR to the corresponding items. Really excited to see what we have built so far!!

startupgurukul commented:
Almost there almost there :) 

Just to let you folks know, the performance of the resnet model is awesome, it's just little underperforming at the palm level and fingers level. 

(There is Ego hands dataset in case if you have plans to improve palm level performance)

tylerzhu-github commented:
> Almost there almost there :)
> 
> Just to let you folks know, the performance of the resnet model is awesome, it's just little underperforming at the palm level and fingers level.
> 
> (There is Ego hands dataset in case if you have plans to improve palm level performance)

Thanks for the interests and testing. I will finish some final polish of the README tomorrow when I am in the office. After that this PR is ready to be merged. Thanks also for the suggestion about palm area segmentation. We will keep maintaining and releasing better quality models after the initial release.

oveddan commented:
Woohoo!  Glad to see this merged!

startupgurukul commented:
Congratulations on commit folks, 

The palm, especially finger  segmentation do you think supporting output stride of 8 for Resnet improve the performance. 

outputStride - Can be one of 8, 16, 32 (Stride 16, 32 are supported for the ResNet architecture and stride 8, 16, 32 are supported for the MobileNetV1 architecture). It specifies the output stride of the BodyPix model.
